srv/modules/runners/rebuild.py

```diff
@@ -0,0 +1,257 @@
+        __master_opts__['quiet'] = True
+        self.qrunner = salt.runner.RunnerClient(__master_opts__)
+        self.minions = self._minions(*args)
+        self.operation = kwargs.get('operation', 'osd.remove')
+        self.aggressive = kwargs.get('aggressive', False)
+        self.skipped = []
+
+    def _minions(self, *args):
+        """ Return expanded Salt target """
+        minions = set()
+        for arg in args:
+            results = self.local.cmd(arg, 'grains.get', ['id'], tgt_type="compound")
+            minions.update(results)
+        log.info("minions: {}".format(list(minions)))
+        return sorted(minions)
+
+    def _osd_list(self, minion):
+        """ Return the list of OSD IDs """
+        osd_ret = self.local.cmd(minion, 'osd.list', [], tgt_type="compound")
+        log.debug("osd.list returned: {}".format(osd_ret))
+        if minion in osd_ret:
+            return osd_ret[minion]
+        return None
+
+    def _validate_osd_df(self, osd_df):
+        """ Validate osd.df output """
+        if self.master_minion not in osd_df:
+            log.error("salt {} osd.df failed".format(self.master_minion))
+            return False
+
+        if 'nodes' not in osd_df[self.master_minion]:
+            log.error("Mangled output: missing 'nodes' from osd.df")
+            return False
+
+        if 'summary' not in osd_df[self.master_minion]:
+            log.error("Mangled output: missing 'summary' from osd.df")
+            return False
+
+        if 'total_kb_avail' not in osd_df[self.master_minion]['summary']:
+            log.error("Mangled output: missing 'total_kb_avail' from osd.df")
+            return False
+        return True
+
+    def safe(self, osds):
+        """ Check OSD used capacity does not exceed available cluster space """
+        osd_df = self.local.cmd(self.master_minion, 'osd.df', [], tgt_type="compound")
+        log.debug("osd_df: {}".format(pprint.pformat(osd_df)))
+        if not self._validate_osd_df(osd_df):
+            return False
+
+        used = 0
+        for osd in osd_df[self.master_minion]['nodes']:
+            if str(osd['id']) in osds:
+                used += osd['kb_used']
+
+        available = osd_df[self.master_minion]['summary']['total_kb_avail']
+        log.info("Used: {} KB  Available: {} KB".format(used, available))
+        if used > available:
+            log.critical("OSDs exceed available free space")
+            return False
+        return True
+
+    def _busy_wait(self):
+        """ Wait until OSDs are safe to stop """
+        delay = 3
+        while True:
+            quiescent = self.local.cmd(self.master_minion,
+                                       'osd.ceph_quiescent', [], tgt_type="compound")
+            log.debug("quiescent: {}".format(quiescent))
+            if self.master_minion in quiescent and quiescent[self.master_minion] is True:
+                break
+            print("Waiting for PGs to recover...")
+            time.sleep(delay)
```

Comment:
This sleeps potentially forever. I think we should add a timeout (a very long though).

Reply:
It does, but the runner is interactive and printing to the terminal.  For the complaints we had in the past regarding any timeout, this would address that.  A Ctrl-C is enough to end it when the admin is tired of waiting and finally decides to investigate.

Reply:
I'm also fine with that